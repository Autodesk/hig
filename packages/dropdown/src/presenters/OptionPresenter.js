import React, { Component } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";
import { ThemeContext } from "@hig/theme-context";
import { CheckmarkSUI, CheckmarkXsUI } from "@hig/icons";
import { createButtonEventHandlers } from "@hig/utils";
import stylesheet from "./OptionPresenter.stylesheet";

function OptionWrapper(props) {
  const {
    children,
    disabled,
    id,
    onClick,
    onMouseDown,
    onMouseMove,
    selected,
    highlighted
  } = props;

  const { handleClick, handleKeyDown } = createButtonEventHandlers(onClick);

  return (
    <ThemeContext.Consumer>
      {({ resolvedRoles }) => (
        <div
          aria-selected={selected}
          className={css(
            stylesheet(
              { disabled, selected, highlighted, ...props },
              resolvedRoles
            ).option
          )}
          disabled={disabled}
          id={id}
          onClick={handleClick}
          onKeyDown={handleKeyDown}
          onMouseDown={onMouseDown}
          onMouseMove={onMouseMove}
          role="option"
          tabIndex="0"
        >
          {children}
        </div>
      )}
    </ThemeContext.Consumer>
  );
}

OptionWrapper.propTypes = {
  children: PropTypes.node,
  disabled: PropTypes.bool,
  highlighted: PropTypes.bool,
  id: PropTypes.string,
  onClick: PropTypes.func,
  onMouseDown: PropTypes.func,
  onMouseMove: PropTypes.func,
  selected: PropTypes.bool
};

export default class OptionPresenter extends Component {
  static propTypes = {
    /**
     * Visual representation of the option
     */
    children: PropTypes.node,
    /**
     * Indicates the option is currently highlighted.
     * This is comparable to hover state, but useful when interacting by keyboard.
     */
    highlighted: PropTypes.bool,
    /**
     * Called when user finishes clicking on an option
     */
    onClick: PropTypes.func,
    /**
     * Called when user begins clicking on an option
     */
    onMouseDown: PropTypes.func,
    /**
     * Called when user moves mouse over the option
     */
    onMouseMove: PropTypes.func,
    /**
     * Indicates the option is currently selected
     */
    selected: PropTypes.bool
  };

  render() {
    const { children, selected, ...otherProps } = this.props;

    return (
      <ThemeContext.Consumer>
        {({ metadata }) => {
          const Icon =
            metadata.densityId === "medium-density"
              ? CheckmarkSUI
              : CheckmarkXsUI;
          return (
            <OptionWrapper selected={selected} {...otherProps}>
              <span>{children}</span>
              {selected && (
                <div>
                  <Icon color="#087AAD" />
                </div>
              )}
            </OptionWrapper>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
