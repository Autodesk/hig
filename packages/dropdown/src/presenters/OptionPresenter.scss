@import "~@hig/styles/mixins/colors.scss";
@import "~@hig/styles/mixins/typography.scss";

.hig__dropdown-option {
  @import "../tokens.scss";
  @include typography-base;

  position: relative;
  display: flex;
  box-sizing: border-box;
  min-height: $option-height;
  align-items: center;
  justify-content: space-between;
  padding: 0 8px;
  cursor: pointer;

  &:hover,
  &.hig__dropdown-option--highlighted,
  &.hig__dropdown-option--focused {
    background-color: color(hig-blue-10);
  }
}
