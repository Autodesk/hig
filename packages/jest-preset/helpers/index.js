/* eslint-disable global-require */
module.exports = {
  takeSnapshotsOf: require("./takeSnapshotsOf")
};
