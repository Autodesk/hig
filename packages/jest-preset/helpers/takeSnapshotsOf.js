const React = require("react");
const renderer = require("react-test-renderer");

module.exports = function takeSnapshotsOf(Component, cases, rendererOptions) {
  cases.forEach(({ desc, props }) => {
    it(desc, () => {
      const vDom = React.createElement(Component, props);
      const tree = renderer.create(vDom, rendererOptions).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
};
