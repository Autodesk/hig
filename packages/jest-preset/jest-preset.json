{
  "collectCoverageFrom": ["**/src/**/*.{js,jsx}"],
  "coverageDirectory": "<rootDir>/coverage",
  "coveragePathIgnorePatterns": [
    "__gemini__",
    "__stories__",
    "docs",
    "examples/redux",
    "node_modules",
    "packages/jest-preset",
    "packages/vanilla",
    "src/playground"
  ],
  "coverageThreshold": {
    "global": {
      "branches": 50,
      "functions": 50,
      "lines": 50,
      "statements": 50
    }
  },
  "moduleNameMapper": {
    "\\.(css|scss|svg)$": "@hig/jest-preset/fileMock.js"
  },
  "modulePaths": ["node_modules"],
  "moduleFileExtensions": ["js", "jsx", "json"],
  "setupFiles": ["raf/polyfill"],
  "setupTestFrameworkScriptFile": "@hig/jest-preset/setupTests.js",
  "testPathIgnorePatterns": [
    "docs",
    "packages/babel-preset",
    "packages/jest-preset",
    "packages/vanilla"
  ],
  "unmockedModulePathPatterns": ["node_modules"]
}
