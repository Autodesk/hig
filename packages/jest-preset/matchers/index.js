require("./toBeConstants");
require("./toImplementHIGInterfaceOf");
require("./toHavePropTypesAndDocgenInfo");
