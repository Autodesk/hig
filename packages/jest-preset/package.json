{
  "private": true,
  "name": "@hig/jest-preset",
  "version": "0.1.0",
  "devDependencies": {
    "babel-core": "^6.26.3",
    "babel-jest": "^23.4.0",
    "enzyme": "^2.9.1",
    "jest": "^21.2.1",
    "jest-enzyme": "^3.8.3",
    "jest-extended": "^0.6.0",
    "raf": "^3.4.0",
    "react-test-renderer": "^15.6.1"
  },
  "eslintConfig": {
    "extends": "@hig"
  },
  "jest": {
    "preset": "@hig/jest-preset"
  }
}
