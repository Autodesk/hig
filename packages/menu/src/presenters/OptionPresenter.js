import React, { Component } from "react";
import PropTypes from "prop-types";
import { css, cx } from "emotion";
import { ThemeContext } from "@hig/theme-context";
import { CheckmarkSUI, CheckmarkXsUI } from "@hig/icons";
import { createCustomClassNames } from "@hig/utils";

import stylesheet from "./stylesheet";
import { AVAILABLE_ROLES } from "../constants";

export default class OptionPresenter extends Component {
  static propTypes = {
    asset: PropTypes.node,
    checkmark: PropTypes.bool,
    children: PropTypes.node,
    disabled: PropTypes.bool,
    highlighted: PropTypes.bool,
    id: PropTypes.string,
    isPressed: PropTypes.bool,
    role: PropTypes.oneOf(AVAILABLE_ROLES),
    selected: PropTypes.bool,
    shortcut: PropTypes.node,
    stylesheet: PropTypes.func
  };

  render() {
    const {
      asset,
      checkmark,
      children,
      disabled,
      highlighted,
      id,
      isPressed,
      role,
      selected,
      shortcut,
      stylesheet: customStylesheet,
      ...otherProps
    } = this.props;
    const { className } = otherProps;
    const menuOptionClassName = createCustomClassNames(
      className,
      `menu-option`
    );
    const checkmarkWrapperClassName = createCustomClassNames(
      className,
      `checkmark-wrapper`
    );
    const assetWrapperClassName = createCustomClassNames(
      className,
      `asset-wrapper`
    );
    const optionContentWrapperClassName = createCustomClassNames(
      className,
      `option-content-wrapper`
    );
    const shortcutWrapperClassName = createCustomClassNames(
      className,
      `shortcut-wrapper`
    );
    const payload = { ...otherProps };
    delete payload.getActiveOption;
    delete payload.getHighlightIndex;
    delete payload.getOptionsInfo;
    delete payload.getPreviousEvent;
    delete payload.setActiveOption;
    delete payload.setHighlightIndex;

    const ariaPayload = role === `option` ? { "aria-selected": selected } : {};

    return (
      <ThemeContext.Consumer>
        {({ resolvedRoles, metadata }) => {
          const styles = stylesheet(
            {
              disabled,
              highlighted,
              isPressed,
              role,
              selected,
              shortcut,
              stylesheet: customStylesheet
            },
            resolvedRoles
          );
          const Checkmark =
            metadata.densityId === `medium-density`
              ? CheckmarkSUI
              : CheckmarkXsUI;
          return (
            <li
              // conditional payload for aria-selected
              {...ariaPayload}
              {...payload}
              className={cx([menuOptionClassName, css(styles.menuOption)])}
              disabled={disabled}
              id={id}
              role={role}
              selected={selected}
            >
              {checkmark && role !== `presentation` ? (
                <div
                  className={cx([
                    checkmarkWrapperClassName,
                    css(styles.checkmarkWrapper)
                  ])}
                >
                  <Checkmark />
                </div>
              ) : null}
              {asset ? (
                <div
                  className={cx([
                    assetWrapperClassName,
                    css(styles.assetWrapper)
                  ])}
                >
                  {asset}
                </div>
              ) : null}
              <div
                className={cx([
                  optionContentWrapperClassName,
                  css(styles.optionContentWrapper)
                ])}
              >
                {children}
                {shortcut ? (
                  <span
                    className={cx([
                      shortcutWrapperClassName,
                      css(styles.shortcutWrapper)
                    ])}
                  >
                    {shortcut}
                  </span>
                ) : null}
              </div>
            </li>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
