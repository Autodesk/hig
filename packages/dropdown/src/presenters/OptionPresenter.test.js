import React from "react";
import { takeSnapshotsOf } from "@hig/jest-preset/helpers";

import OptionPresenter from "./OptionPresenter";

describe("Dropdown/presenters/OptionPresenter", () => {
  const rendererOptions = {
    createNodeMock(element) {
      return document.createElement(element.type);
    }
  };
  const cases = [
    {
      desc: "renders without props",
      props: {}
    },
    {
      desc: "renders with all props",
      props: {
        highlighted: true,
        onClick: () => {},
        onMouseDown: () => {},
        onMouseMove: () => {},
        onHover: () => {},
        onSelect: () => {},
        selected: true,
        value: "my value",
        children: <p key="p">hi</p>
      }
    }
  ];

  takeSnapshotsOf(OptionPresenter, cases, rendererOptions);
});
