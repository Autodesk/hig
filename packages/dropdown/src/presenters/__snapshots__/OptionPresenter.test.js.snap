// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown/presenters/OptionPresenter renders with all props 1`] = `
.emotion-1 {
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  box-sizing: border-box;
  min-height: 36px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 8px;
  cursor: pointer;
  color: #3c3c3c;
  font-family: ArtifaktElement,sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 1.428571429;
  background-color: rgba(128,128,128,0.08);
}

.emotion-0 {
  fill: #087AAD;
  pointer-events: none;
}

.emotion-0 > * {
  fill: #087AAD;
  pointer-events: none;
}

<div
  aria-selected={true}
  className="emotion-1"
  disabled={undefined}
  id={undefined}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseMove={[Function]}
  role="option"
  tabIndex="0"
>
  <span>
    <p>
      hi
    </p>
  </span>
  <div>
    <svg
      className="emotion-0"
      color="#087AAD"
      height="16px"
      viewBox="0 0 16 16"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        checkmark-s
      </title>
      <path
        d="M6.2 12L2 8.1l1.3-1.3 2.9 2.7L12.7 3 14 4.3 6.2 12z"
      />
    </svg>
  </div>
</div>
`;

exports[`Dropdown/presenters/OptionPresenter renders without props 1`] = `
.emotion-0 {
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  box-sizing: border-box;
  min-height: 36px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 8px;
  cursor: pointer;
  color: #3c3c3c;
  font-family: ArtifaktElement,sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 1.428571429;
  background-color: transparent;
}

<div
  aria-selected={undefined}
  className="emotion-0"
  disabled={undefined}
  id={undefined}
  onClick={undefined}
  onKeyDown={undefined}
  onMouseDown={undefined}
  onMouseMove={undefined}
  role="option"
  tabIndex="0"
>
  <span />
</div>
`;
